Add Support for Tracking Parameter Names (named_parameters) in Optimizer State Dict

ErezYosef opened the pull request:
A proposal addressing Issue #1489: **Optimizer should track parameter names and not id.**

(also mentioned in here: [[RFC] Introducing FQNs/clarity eyeglasses to optim state_dict](https://dev-discuss.pytorch.org/t/rfc-introducing-fqns-clarity-to-optim-state-dict/1552)

## Summary
This PR introduces a backward-compatible enhancement where optimizers track parameter names instead of just their id. 
Optimizers can be initialized with `named_parameters()` as:
```python
optimizer = optim.SGD(model.named_parameters(), lr=0.01, momentum=0.9)
```
This allows for greater clarity and ease when handling optimizers, as the parameters' names are preserved within the optimizer’s `state_dict` as: 
```
state_dict = 
{
    'state': {
    0: {'momentum_buffer': tensor(...), ...},
    1: {'momentum_buffer': tensor(...), ...},
    },
    'param_groups': [
        {
        'lr': 0.01,
        'weight_decay': 0,
        ...
        'params': [0,1]
        'param_names' ['layer.weight', 'layer.bias']  (optional)
        }
    ]
}
```
Loading `state_dict` is not changed (backward-compatible) and the `param_names` key will be ignored.


## Key Features
#### Named Parameters in Optimizer Initialization:
Optimizers can accept the output of `model.named_parameters()` during initialization, allowing them to store parameter names directly.
#### Parameter Names in `state_dict`:
The parameter names are saved as a list in the optimizer’s `state_dict` with key `param_names`, alongside the `params` indices, ensuring seamless tracking of both names and parameters.

## Backward Compatibility
#### No Breaking Changes:
This change is fully backward-compatible. The added `param_names` key in the optimizer's `state_dict` is ignored when loading a state to the optimizer.

#### Customization with Hooks:
For more control, the loaded state_dict can be modified using a custom `register_load_state_dict_pre_hook`, providing flexibility for different design needs.

## Documentation Updates
Please refer to the documentation changes for more details on how this feature is implemented and how it can be used effectively.

## Solution Example:

A suggested solution to the problem mentioned in #1489, for the same parameters but in a different order.
The following `register_load_state_dict_pre_hook` should be added to the optimizer before loading to enable loading the state dict :
```python
def adapt_state_dict_ids(optimizer, state_dict):
    # assuming a single param group. 
    current_state_group = optimizer.state_dict()['param_groups'][0]
    loaded_state_group = state_dict['param_groups'][0]

    # same number of params, same names, only different ordering
    current_state_name_to_id_mapping = {}  # mapping --  param_name: id
    for i, name in enumerate(current_state_group['param_names']):
        current_state_name_to_id_mapping[name] = current_state_group['params'][i]

    # changing the ids of the loaded state dict to match the order of the given state dict.
    for i, name in enumerate(current_state_group['param_names']):
        loaded_state_group['params'][i] = current_state_name_to_id_mapping[name]

    return state_dict
```
In this code, the loaded `state_dict` ids are adapted to match the order of the current optimizer `state_dict`.
Both the previous and the current optimizers are required to be initiated with `named_parameters()` to have the 'param_names' key in the dict.

### Note
This is my first contribution to PyTorch, and I wish to receive feedback or suggestions for improvement.

ErezYosef committed:
Adding support for named_parameters in optimizer state dict.

vadimkantorov commented:
Is this related? 
- https://github.com/pytorch/pytorch/issues/72146

ErezYosef commented:
> Is this related?
> 
> * [Also allow dicts as type of `params=` field in param groups of optimizers #72146](https://github.com/pytorch/pytorch/issues/72146)

Issue #72146 mentions this problem of missing names, but suggests changing the `state_dict['param_groups']['params']` from list to dict. It is an option but backward compatibility and the ordering of the parameters should be checked.

Also this problem is mentioned in Issue 71683 [comment 1019456452](https://github.com/pytorch/pytorch/issues/71683#issuecomment-1019456452) as said: _"provide param_names list inside the param group"_ , as suggested here.

janeyx99 commented:
hi @ErezYosef! Thanks for the well-written proposal and implementation attempt. I agree this is a good direction to go towards, where we gently enable accepting named_parameters without breaking BC.

Before I review your PR in full, it'd be essential to add testing and docs:
- Add tests in test/test_optim.py to verify behavior. For example, I would expect that an optimizer given the named_parameters would be saved and checkpointed properly. We'd also want to make sure that nothing semantically changes for all our implementations when parameters are named_parameters().
- Add documentation for each optimizer in their respective files (though should mostly be modularized in optimizer.py).
- Add docs to optim.rst to show the use case people would use this for

ErezYosef committed:
Adding docs and tests.

ErezYosef commented:
Thanks @janeyx99 .
I pushed the changes you requested.
Please review the test I added, I hope it is good.

Erez.

vadimkantorov commented:
Maybe also this is related:
- https://github.com/pytorch/pytorch/issues/104247
could be an alternative, e.g. when `.parameters()` is called it would e.g. it could return `param.with_name(name)` which could then be used

With `named_parameters()` support of optimizers, maybe it's less pressing

ErezYosef commented:
> Maybe also this is related:
> 
> * [[proposal] "Name" string attribute for modules, parameters, buffers, tensors for more pleasant debugging (especially for graph printouts / export / studying compiled generated code) #104247](https://github.com/pytorch/pytorch/issues/104247)
> 
> could be an alternative, e.g. when `.parameters()` is called it would e.g. it could return `param.with_name(name)` which could then be used
> 
> With `named_parameters()` support of optimizers, maybe it's less pressing

 It looks like a very useful feature that could enhance debugging and the overall experience.
Once it is fully characterized and implemented the optimizers can also support it (according to the specific design choices that need to be made).

However, the current PR contributes to the existing code as it is, and enables more flexibility (without having major changes).
Therefore, I believe it can be merged independently of any potential future features.

Erez.

janeyx99 commented on test/test_optim.py:
```diff
@@ -1462,6 +1462,67 @@ def fwd_bwd(optim, mod, i):
                 optimizer.step()
 
+    @optims(optim_db, dtypes=[torch.float32])
+    def test_can_load_from_to_named_state_dict(self, device, dtype, optim_info):
```
Briefly looked over this test -- it is good to test this.

Overall, I am expecting more tests where someone passes in named_parameters to the constructor, and they should work for all our optimizers.

janeyx99 replied:
I'm expecting a test outside of this test case that shows that for every config, passing named_parameters vs passing just normal parameters yields the same results

ErezYosef committed:
Adding BC tests: same results with named_parameters, and satet_dict save-load equality

ErezYosef committed:
Adding docs explanations to utilize the new feature.

ErezYosef committed:
fix example typos

ErezYosef commented:
Hi @janeyx99 

I added a few more tests + explanations and an example to present how to use the new feature.
Please check these tests and verify their correctness.

Thanks, 
Erez.

janeyx99 commented on docs/source/optim.rst:
```diff
@@ -303,6 +314,76 @@ algorithms.
+--------------------------------------------
+
+The function :func:`~Optimizer.load_state_dict` ignores the optional ``param_names`` input in the
+loaded state dict if present (to maintain compatibility).
```
Is it better to say it treats like any other unfamiliar key in the state and will populate by just copy/paste?

janeyx99 commented on docs/source/optim.rst:
```diff
@@ -303,6 +314,76 @@ algorithms.
+    optimizer = optim.SGD(model.named_parameters(), lr=0.01, momentum=0.9)
+    torch.save(optimizer.state_dict(), PATH)
+
+Assuming that for the following ``model2`` we wish to load the states of ``fc`` of ``model`` into both ``fc1`` and ``fc2`` of ``model2``::
```
Do you have a concrete example of when someone wants to do this?

Higher level question: I would imagine a better example for registering a custom hook is to maintain the same names as the previous state_dict. Is that already the default behavior with no hook --> is there a test below that tests this?

Or you can answer this q: what happens without the custom hook? Will the param_names get copied over or will they get completely ignored (the new state_dict no longer has them)?

janeyx99 commented on test/test_optim.py:
```diff
@@ -1368,51 +1378,55 @@ def fwd_bwd(optim, w, b, i):
-                else:
-                    closure()
-                    optimizer.step()
+            for param_in in make_param_and_named_param(params):
```
Stylistic nit: Instead of parametrizing in here, could you add this as a parameterization to the test (so there are two distinct tests vs one large test)?

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -1018,6 +1031,27 @@ def add_param_group(self, param_group: Dict[str, Any]) -> None:
+                    param_group['param_names'] = []
+                    extracted_param_tensors = []
+                    warnings.warn(
+                        "not all parameters in optimizer's parameter group are named_parameters. name is missing."
```
I'm inclined to not want to allow this case at all--in other words, I think we should just error without affecting the optimizer state.

janeyx99 commented on test/test_optim.py:
```diff
@@ -1462,6 +1476,67 @@ def fwd_bwd(optim, mod, i):
+                        optimizer2.step(functools.partial(fwd_bwd, optimizer2, model, input))
+                    else:
+                        fwd_bwd(optimizer2, model, input)
+                        optimizer2.step()
```
Following comments from above, could we add another test like this one except just showing what the behavior should be? i.e., are the names preserved?

ErezYosef replied:
Tested in L1536:
``` python
            # Make sure that param_names are preserved when provided to at least one of the optimizers
            if is_named_optim0 or is_named_optim1:
                self.assertEqual(optimizer2.state_dict()['param_groups'][0]['param_names'],
                                 ['0.weight', '0.bias', '1.weight', '1.bias'])
```

janeyx99 commented on test/test_optim.py:
Instead of creating more nested lines in each of the tests below, could you

1. use our @parametrize (examples throughout the file) for keeping the test cleaner
2. add a test similar to:
```
    @optims([o for o in optim_db if not o.only_supports_sparse_grads], dtypes=[torch.float32])
    def test_named_params_same_behavior(self, device, dtype, optim_info):
        optim_cls = optim_info.optim_cls
        # Skip differentiable testing for now, see https://github.com/pytorch/pytorch/issues/116490
        all_optim_inputs = _get_optim_inputs_including_global_cliquey_kwargs(
            device, dtype, optim_info, skip=("differentiable",)
        )
        for optim_input in all_optim_inputs:
            torch.manual_seed(1)
			model = torch.nn.Sequential(
                torch.nn.Linear(2, 3),
                torch.nn.Sigmoid(),
                torch.nn.Linear(3, 1),
                torch.nn.Sigmoid(),
            )

            torch.manual_seed(1)
			model2 = torch.nn.Sequential(
                torch.nn.Linear(2, 3),
                torch.nn.Sigmoid(),
                torch.nn.Linear(3, 1),
                torch.nn.Sigmoid(),
            )
            model1.to(dtype=dtype, device=device)
            model2.to(dtype=dtype, device=device)

            optimizer = optim_cls(model1.parameters(), **optim_input.kwargs)
            optimizer_named = optim_cls(model2.named_parameters(), **optim_input.kwargs)

            loss = (weight.mv(input) + bias).pow(2).sum()
            initial_value = loss.item()
            input = torch.tensor(
                    [0.1, 0.2, 0.3, 0.4, 0.5, 0.6], dtype=dtype, device=device
                ).reshape(3, 2)

            for _ in range(3):
                optimizer.zero_grad()
                optimizer_named.zero_grad()
                loss1 = model1(input)
			    loss2 = model2(input)
                loss1.backward()
                loss2.backward()
                optimizer.step()
                optimizer_named.step()
                for ref, actual in zip(model1.parameters(), model2.parameters()):
                		self.assertEqual(ref, actual)
```

I just wrote that test on GitHub UI, but that'd be the essence!

ErezYosef replied:
I assume the request:

``` python
                for ref, actual in zip(model1.parameters(), model2.parameters()):
                		self.assertEqual(ref, actual)

```
is already satisfied by lines 1416-1417:

``` python
                self.assertEqual(weight, weight_c)
                self.assertEqual(bias, bias_c)

```

janeyx99 reviewed (commented):
Sorry this review took me so long--I just got back from hyperfocusing on the conference + traveling to the conference and am prioritizing responding to GH notifications now! I would like to get your PR in by the end of this week if possible.

Thanks for your work and for adding tests--I've added some more highlevel comments throughout. Mainly, I'm curious how far we want to support:
1. Do we want to support mixed names + no names within a param group? (I'm thinking no)
2. Do we want to support param groups with names + param groups without names within one optimizer? (Maybe, but if there's no real use case for this, I'd also argue to stay as simple as we can).
3. What would you guess is the main use case for using named parameters? Is it for easier filtering later? For debugging? For sharding the statedict? Whatever that is, I'd like to showcase an example in the docs.
4. What is the expected behavior of: 
```
optim = AdamW(model.named_parameters())
named_sd = optim.state_dict()

optim2 = AdamW(model.parameters())
optim2.load_state_dict(named_sd)
```
and
```
optim = AdamW(model.parameters())
unnamed_sd = optim.state_dict()

optim2 = AdamW(model.named_parameters())
optim2.load_state_dict(unnamed_sd)
```
5. For each of the above, I'd love to get a simple test case going exhibiting the behavior. For example, if we want to error, we should add combinations of mixed names/no names in get_error_inputs_for_all_optims.

ErezYosef commented:
Thanks so much for your detailed review, @janeyx99, and no worries about the timing! 

Regarding your questions:

1. I completely agree with you. I’ll ensure that we don’t mix parameters with and without names within a parameter group.
2. Keeping things simple sounds like a good approach, and I can enforce consistent behavior across all groups. However, this would introduce a constraint that may not be strictly necessary.
3. Named parameters primarily offer flexibility for various use cases, including easier filtering, debugging, and managing the state_dict, as you mentioned. I’ll add examples to showcase how this can be useful. 
The example I added in `optim.rst` (involving fc1 and fc2) is similar to the scenario that led me to open this PR (training of multi-experts).
It can also be used to apply something like "strict=False" #3852  (see added example)
4. The behavior in these cases would ignore the names in the state dict (for backward compatibility), treating them as the optimizer was initialized with model.parameters(). I’ll add tests to cover this as you suggested.
5. 
I’ll make sure to include test cases to capture these behaviors, including scenarios where we expect an error if mixed names are used.

Thanks again,
Erez.

ErezYosef committed:
code adaptations and docs examples

ErezYosef committed:
adding the required tests

janeyx99 commented:
Ah thank you. Let me know when this PR is ready for another review--I've kicked off the CI for now.

ErezYosef commented:
> Ah thank you. Let me know when this PR is ready for another review--I've kicked off the CI for now.

Thanks,
I assume these commits address all your points.

Addressing the points:
1. I added the raise:
``` python 
        if len(extracted_param_names) != 0:
            if len(extracted_param_names) == len(extracted_param_tensors):
                param_group['param_names'] = extracted_param_names
            else:
                raise ValueError("all optimizer params should be with/without names. Some param names are missing")

```

2. I added:
``` python 
                ....
                raise ValueError("all optimizer param groups should be with/without names. "
                                 f'cannot add param group {current_group_txt} to the optimizer')

```

3. Examples were added to the doc
4. I add a test for the behavior:
``` python 
            # Make sure that param_names are preserved when provided to at least one of the optimizers
            if is_named_optim0 or is_named_optim1:
                self.assertEqual(optimizer2.state_dict()['param_groups'][0]['param_names'],
                                 ['0.weight', '0.bias', '1.weight', '1.bias'])
```

Feel free to let me know if there's anything else to add.
Erez.

janeyx99 commented on test/test_optim.py:
```diff
@@ -1462,8 +1478,69 @@ def fwd_bwd(optim, mod, i):
+
+            # old_state_dict has all new flags del'd
+            old_state_dict = deepcopy(optimizer.state_dict())
+            old_state_dict_pg = old_state_dict["param_groups"]
+            for group in old_state_dict_pg:
+                for flag in optim_info.not_og_supported_flags:
+                    if flag in group:
+                        del group[flag]
```
is this code needed in this test? Looks like remains of copypasta 😛

ErezYosef replied:
Thanks, I wasn't sure about that, so I'll go ahead and delete it.

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -653,6 +657,8 @@ def state_dict(self) -> StateDict:
             parameter group is a Dict. Each parameter group contains metadata
             specific to the optimizer, such as learning rate and weight decay,
             as well as a List of parameter IDs of the parameters in the group.
+            if a param group was initiated with ``named_parameters()`` the names
```
```suggestion
            If a param group was initialized with ``named_parameters()`` the names
```

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -834,6 +842,14 @@ def register_load_state_dict_post_hook(
     def load_state_dict(self, state_dict: StateDict) -> None:
         r"""Load the optimizer state.
+        The names of the parameters (if they exist in :meth:`state_dict`) will not affect the loading process.
+        To use the parameters names for custom cases (such as when the parameters in the loaded state dict
```
```suggestion
        To use the parameters' names for custom cases (such as when the parameters in the loaded state dict
```

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -834,6 +842,14 @@ def register_load_state_dict_post_hook(
     @torch._disable_dynamo
     def load_state_dict(self, state_dict: StateDict) -> None:
         r"""Load the optimizer state.
+        The names of the parameters (if they exist in :meth:`state_dict`) will not affect the loading process.
```
maybe specify `under the "param_names" key of each param group`? to be precise

janeyx99 replied:
Put this entire explanation (whcih is very good btw!) under a Note so people know it's not the main carb of the function

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -834,6 +842,14 @@ def register_load_state_dict_post_hook(
+         differ from those initialized in the optimizer),
+        a custom ``register_load_state_dict_pre_hook`` should be implemented to adapt the loaded dict
+         accordingly.
+         If ``param_names`` exist in loaded state dict ``param_groups`` they will be saved or will override
```
```suggestion
         If ``param_names`` exist in loaded state dict ``param_groups`` they will be saved and override
```

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -909,6 +925,8 @@ def update_group(
         ) -> Dict[str, Any]:
             new_group["params"] = group["params"]
+            if 'param_names' in group and 'param_names' not in new_group:
+                new_group["param_names"] = group["param_names"]
```
Wait I might be confused--but doesn't this contradict the earlier statement about overriding? This looks like the new_group will only get written IF not previously defined? aka it will not be overridden?

ErezYosef replied:
Note that new_group is the saved_group (from the loaded state_dict).

if `param_names` in both group and new_group: keep (and set) the content of the `new_group["param_names"]` (override previous content). Thus, no changes are needed.
if `param_names` in new_group only: keep the content of the `new_group["param_names"]` (also no changes are needed)
if `param_names` in `group` **only** (this case, see code below): copy the content of `group["param_names"]` to `new_group["param_names"]` which will be set to the optimizer to preserve the names.

In this case, e.g.:
``` python
optimizer_named = optim_cls(model2.named_parameters(), **optim_input.kwargs)
optimizer_named .load_state_dict(state_dict_without_names)
```

This behavior is important for addressing situations where you have a `non_named_state_dict` and want to obtain a `named_optimizer` or `named_state_dict`. This is achieved by loading the `non_named_state_dict` into a named optimizer, and when saving its `state_dict`, the parameter names will be included.

Please let me know if I've missed anything.

janeyx99 replied:
The new_group being from the saved state_dict is a great distinction. thanks for clarifying

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -1018,6 +1036,23 @@ def add_param_group(self, param_group: Dict[str, Any]) -> None:
+            if len(extracted_param_names) == len(extracted_param_tensors):
+                param_group['param_names'] = extracted_param_names
+            else:
+                raise ValueError("all optimizer params should be with/without names. Some param names are missing")
```
nice

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -1049,6 +1084,10 @@ def add_param_group(self, param_group: Dict[str, Any]) -> None:
             param_set.update(set(group["params"]))
+            if ('param_names' in param_group) != ('param_names' in group):
+                current_group_txt = 'with names' if 'param_names' in param_group else 'without names'
+                raise ValueError("all optimizer param groups should be with/without names. "
```
yea i agree with you this might be an unnecessary constraint but if someone wants this feature later it shouldn't be awful to add it. thanks for the error here!

janeyx99 reviewed (commented):
Looks really good! CI looks like a mess though--could you rebase and let's see if these tests all pass.

ErezYosef commented on docs/source/optim.rst:
```diff
@@ -303,6 +314,172 @@ algorithms.
+loaded state dict if present. However, the process of loading the optimizer state is not affected,
+as the order of the parameters matters to maintain compatibility (in case of different ordering).
+To utilize the loaded parameters names from the loaded state dict, a custom ``register_load_state_dict_pre_hook``
+needs to be implemented according to the desired behavior.
```
Yes, load_state_dict (named_state_dict) will not base any decisions on the names (also for BC !)
All changes from the predicted (existing) behavior should be manually set by adding a "hook" to the process.

** This is beneficial since this feature is usually needed when there is a mismatch in the state dict (such as different layers, different order, etc.) and some manual adaptations are required according to the needs

janeyx99 replied:
oop yea i...forgot to come back to this comment once i had read the rest of the code.

ErezYosef commented:
Thanks.
I am not sure how the CI works. Hope it will work easily after a rebase.

Do you want me to rebase on upstream viable/strict? or which branch?
should it be forced?

Do you know if these commands will work?
```
git pull --rebase upstream viable/strict
git push -f
```

Erez.

janeyx99 commented:
Rebasing onto viable/strict or main would both work. the commands you wrote should be fine—i typically go for a git fetch followed by a git rebase.

ErezYosef committed:
Fixes and improvements based on review feedback

ErezYosef commented:
Hi @janeyx99 .

Please review my last commit and rebase.
Hope it will be ok now.

Erez.

janeyx99 commented on docs/source/optim.rst:
```diff
@@ -303,6 +314,185 @@ algorithms.
+How to utilize named parameters to load optimizer state dict
+------------------------------------------------------------
+
+The function :func:`~Optimizer.load_state_dict` store the optional ``param_names``content in the
```
```suggestion
The function :func:`~Optimizer.load_state_dict` stores the optional ``param_names``content from the
```

janeyx99 commented on docs/source/optim.rst:
```diff
@@ -303,6 +314,185 @@ algorithms.
+    # training..
+    torch.save(optimizer.state_dict(), PATH)
+
+Assuming that ``model`` implements an expert (MoE) and we want to duplicate it and resume training
+for two experts, both initialized same as the ``fc`` layer.
+For the following ``model2`` we create two layers identical to ``fc`` and resume training by loading the model weights
+and optimizer states from ``model`` into both ``fc1`` and ``fc2`` of ``model2``
+(and adjust them accordingly)::
```
```suggestion
Let's say that ``model`` implements an expert (MoE), and we want to duplicate it and resume training
for two experts, both initialized the same way as the ``fc`` layer. For the following ``model2`` wecreate two layers identical to ``fc`` and resume training by loading the model weights and optimizer states from ``model`` into both ``fc1`` and ``fc2`` of ``model2`` (and adjust them accordingly)::
```

janeyx99 commented on docs/source/optim.rst:
```diff
@@ -303,6 +314,185 @@ algorithms.
+
+To load the state dict for ``optimizer2`` with the state dict of the previous optimizer such that both
+``fc1`` and ``fc2`` will be initialized with a copy of ``fc`` optimizer states
+(to resume training for each layer from ``fc``), we need the following hook::
```
```suggestion
(to resume training for each layer from ``fc``), we can use the following hook::
```

janeyx99 commented on docs/source/optim.rst:
```diff
@@ -303,6 +314,185 @@ algorithms.
+
+
+
+Instead of loading a state according to the order of parameters (the default approach),
+this hook can be used to load according to the parameters' names::
```
```suggestion
As a third example, instead of loading a state according to the order of parameters (the default approach),
this hook can be used to load according to the parameters' names::
```

janeyx99 reviewed (approved):
Approval contingent on green CI :D (and that you address the nits on wording)

ErezYosef committed:
Apply suggestions from code review

Co-authored-by: Jane (Yuan) Xu <31798555+janeyx99@users.noreply.github.com>

ErezYosef committed:
Merge branch 'main' into optimizer_named_nightly

janeyx99 commented on docs/source/optim.rst:
```diff
@@ -341,11 +341,8 @@ Example::
-and optimizer states from ``model`` into both ``fc1`` and ``fc2`` of ``model2``
-(and adjust them accordingly)::
+Let's say that ``model`` implements an expert (MoE), and we want to duplicate it and resume training
+for two experts, both initialized the same way as the ``fc`` layer. For the following ``model2`` wecreate two layers identical to ``fc`` and resume training by loading the model weights and optimizer states from ``model`` into both ``fc1`` and ``fc2`` of ``model2`` (and adjust them accordingly)::
```
```suggestion
for two experts, both initialized the same way as the ``fc`` layer. For the following ``model2`` we create two layers identical to ``fc`` and resume training by loading the model weights and optimizer states from ``model`` into both ``fc1`` and ``fc2`` of ``model2`` (and adjust them accordingly)::
```

ErezYosef replied:
Fixed in commit 9d7a208

janeyx99 commented:
CI failures are real! [hud.pytorch.org/pr/134107](https://hud.pytorch.org/pr/134107). Please verify the tests pass locally: `python test/test_optim.py -k <test_name>` before committing. lmk if you're stuck on any errors, the SparseAdam ones you can follow the example from the other tests to make the grads sparse 😛

ErezYosef commented:
Thansk.

1. I can see how to make grad sparse. but it is a kind of copy-paste of `test_can_load_older_state_dict` so why is it works on the original test?

2. Not sure I understand this error: 
```
X linux-jammy-py3.9-gcc11 / test (docs_test, 1, 1, linux.2xlarge)
    Queued: 0.0s, Duration: 10.0m, Test insights, Show artifacts, Raw logs

▼ /opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/jit/_trace.py:791: TracerWarning: Output nr 1. of the traced function does not match the corresponding output of the Python function. Detailed error:

/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/jit/_trace.py:791: TracerWarning: 
Output nr 1. of the traced function does not match the corresponding output of the Python function. Detailed error:

2024-10-09T16:01:47.8645453Z Mismatched elements: 4 / 12 (33.3%)
2024-10-09T16:01:47.8646178Z Greatest absolute difference: 0.39807265996932983 at index (0, 2) (up to 1e-05 allowed)
2024-10-09T16:01:47.8647146Z Greatest relative difference: 1.0307998656385509 at index (0, 1) (up to 1e-05 allowed)
2024-10-09T16:01:47.8647768Z   _check_trace(
2024-10-09T16:01:47.8679892Z /opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/jit/_trace.py:791: TracerWarning: Trace had nondeterministic nodes. Did you forget call .eval() on your model? Nodes:
2024-10-09T16:01:47.8681483Z 	%14 : Float(1, 4, strides=[4, 1], requires_grad=0, device=cpu) = aten::rand(%9, %10, %11, %12, %13) # <doctest default[0]>:2:0
2024-10-09T16:01:47.8682670Z This may cause errors in trace checking. To disable trace checking, pass check_trace=False to torch.jit.trace()
2024-10-09T16:01:47.8683459Z   _check_trace(
2024-10-09T16:01:47.8685149Z /opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/jit/_trace.py:791: TracerWarning: Output nr 1. of the traced function does not match the corresponding output of the Python function. Detailed error:
2024-10-09T16:01:47.8686536Z Tensor-likes are not close!

```
What is its source and maybe you have a clue how to solve it?

Thanks, 
Erez.

janeyx99 commented:
Ah, it's skipped in our skip list: https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_optimizers.py#L2163-L2169 I think skipping it is fine as well, though I'd prefer it be in the test case, e.g.,
```
@optims([o for o in optims_db if not o.only_supports_sparse_grads])
def test...
```

janeyx99 commented:
For the docs test, the error in jit is a red herring. The more relevant line is
```
2024-10-09T16:18:53.1562928Z /opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/optim/optimizer.py:docstring of torch.optim.optimizer.Optimizer.load_state_dict:7: WARNING: Block quote ends without a blank line; unexpected unindent.
```

ErezYosef committed:
Lint tests fix

ErezYosef committed:
Tests fix

ErezYosef committed:
Lint minor fix

ErezYosef committed:
test_fix

ErezYosef committed:
fixing indentation trying to make CI satisfied

janeyx99 commented:
For lint, I'd recommend `pip install lintrunner` and then running `lintrunner --init; lintrunner -a` to get all of lint to be fixed for ci.

ErezYosef committed:
lint fix

ErezYosef commented:
> For lint, I'd recommend `pip install lintrunner` and then running `lintrunner --init; lintrunner -a` to get all of lint to be fixed for ci.

Thanks, @janeyx99. However, I'm running into some issues with this command, possibly related to its dependencies.

ErezYosef commented:
Looks like we have green CI. 💯 
What should I do next?

janeyx99 commented:
Wonderful!

@pytorchbot merge

ErezYosef commented on docs/source/optim.rst:
```diff
@@ -303,6 +314,182 @@ algorithms.
+How to utilize named parameters to load optimizer state dict
+------------------------------------------------------------
+
+The function :func:`~Optimizer.load_state_dict` stores the optional ``param_names``content from the
```
Probably missing whitespace after ``param_names``

janeyx99 replied:
if u want to open another pr to address, i'd approve it
